charms/worker/k8s/src/certificates.py

```diff
@@ -0,0 +1,280 @@
+                subject=f"system:node:{node_name}",
+                organization="system:nodes",
+            )
+
+            content["apiserver-kubelet-client-csr"] = apiserver_client.decode()
+            log.info("API Server Client CSR generated")
+
+        self.charm.unit.add_secret(content=content, label=f"{self.charm.unit.name}-csr")
+
+    def _generate_certificate(self, components: List[str]):
+        """Request the creation of certificates for components lacking a certificate.
+
+        Args:
+            components (List[str]): A list of components to request certificates for.
+        """
+        csr = self.charm.model.get_secret(label=f"{self.charm.unit.name}-csr")
+        csr_content = csr.get_content(refresh=True)
+
+        for component in components:
+            if not self._check_secret(f"{component}-cert-{self.charm.unit.name}"):
+                self.certificates.request_certificate_creation(
+                    certificate_signing_request=csr_content[f"{component}-csr"].encode()
+                )
+                log.info(f"{component} Certificate Requested")
+
+    def collect_certificate(self, event: CertificateAvailableEvent):
+        """Collect and store a certificate.
+
+        This method handles the CertificateAvailableEvent by verifying if the
+        CSR from the event matches any stored CSRs. If a match is found, it
+        updates or creates a secret with the certificate, CA certificate, and
+        the CSR.
+
+        Args:
+            event (CertificateAvailableEvent): An event that triggers the certificate
+                                              collection, containing the certificate,
+                                              CA certificate, and CSR.
+        """
+        if not isinstance(event, CertificateAvailableEvent):
+            return
+
+        if not self._check_secret(f"{self.charm.unit.name}-csr"):
+            return
+
+        csr = self.charm.model.get_secret(label=f"{self.charm.unit.name}-csr")
+        csr_content = csr.get_content(refresh=True)
+        csr_keys = set(csr_content.keys())
+
+        for csr in csr_keys:
+            if csr_content[csr].strip() != event.certificate_signing_request:
+                continue
+            certificate_content = {
+                "certificate": event.certificate,
+                "ca-certificate": event.ca,
+                "csr": event.certificate_signing_request,
+            }
+            name = csr.replace("-csr", "")
+            if self._check_secret(label=f"{name}-cert-{self.charm.unit.name}"):
+                certificate_secret = self.charm.model.get_secret(
+                    label=f"{name}-cert-{self.charm.unit.name}"
+                )
+                certificate_secret.set_content(content=certificate_content)
+            else:
+                self.charm.unit.add_secret(
+                    content=certificate_content, label=f"{name}-cert-{self.charm.unit.name}"
+                )
```

Comment:
Surely this certificate expires?  Do you need to read the expiry on the certificate and expire the secret at the same time.  How do you handle regenerating secrets?  Is there someway to test the certificate rotation by changing the expiration time to something super short?

Reply:
That's correct, the library emits events when the certificates are near to the expiry date. The certificate expiration should be implemented in another PR. We still need to implement the mechanism into the snap to allow updating the certificate bits in the cluster. I've created KU-776 for this purpose.